lib/object_pool.js

```diff
@@ -0,0 +1,169 @@
+
+'use strict';
+
+var assert = require('assert');
+
+var DEFAULT_MAX_SIZE = 1000;
+
+module.exports = ObjectPool;
+
+function ObjectPool(options) {
+    assert(typeof options === 'object', 'expected options object');
+
+    this.Type = options.Type;
+    assert(typeof this.Type === 'function', 'expected options.Type to be constructor function');
+    assert(typeof this.Type.prototype.reset === 'function', 'expected options.Type to have reset method');
+    assert(typeof this.Type.prototype.clear === 'function', 'expected options.Type to have clear method');
+
+    this.maxSize = options.maxSize || DEFAULT_MAX_SIZE;
+    assert(typeof this.maxSize === 'number', 'expected options.maxSize to be number');
+
+    this.name = options.name || this.Type.name;
+    assert(typeof this.name === 'string', 'expected options.name to be string');
+
+    this.freeList = [];
+    this.outstanding = 0;
+    this.freeStatName = 'object-pools.' + this.name + '.free';
+    this.outstandingStatName = 'object-pools.' + this.name + '.outstanding';
+}
+
+ObjectPool.statsd = null;
+ObjectPool.reportInterval = null;
+ObjectPool.timers = null;
+ObjectPool.pools = [];
+ObjectPool.timer = null;
+ObjectPool.refs = 0;
+
+ObjectPool.setup = function setup(options) {
+    var pool = new ObjectPool(options);
+    options.Type.alloc = alloc;
+
+    options.Type.prototype.free = function freeThisObj() {
+        pool.free(this);
+    };
+
+    ObjectPool.pools.push(pool);
+
+    return pool;
+
+    function alloc() {
+        var obj = pool.get();
+        return obj;
+    }
+};
+
+ObjectPool.bootstrap = function bootstrap(options) {
+    if (ObjectPool.refs >= 1) {
+        ObjectPool.refs += 1;
+        return;
+    }
+
+    ObjectPool.refs += 1;
+
+    assert(typeof options === 'object', 'expected options object');
+
+    assert(
+        typeof options.statsd === 'object' &&
+        typeof options.statsd.gauge === 'function',
+        'expected options.statsd to be statsd instance'
+    );
+    ObjectPool.statsd = options.statsd;
+
+    assert(
+        typeof options.reportInterval === 'number',
+        'expected options.reportInterval to be number'
+    );
+    ObjectPool.reportInterval = options.reportInterval;
+
+    assert(
+        typeof options.timers === 'object' &&
+        typeof options.timers.setTimeout === 'function',
+        'expected options.timers to be timers object'
+    );
+    ObjectPool.timers = options.timers;
+
+    ObjectPool.timer = ObjectPool.timers.setTimeout(
+        ObjectPool.reportStats,
+        ObjectPool.reportInterval
+    );
+};
+
+ObjectPool.unref = function unref() {
+    ObjectPool.refs = Math.max(0, ObjectPool.refs - 1);
+    if (ObjectPool.refs === 0) {
+        ObjectPool.timers.clearTimeout(ObjectPool.timer);
+        ObjectPool.timer = null;
+    }
+};
+
+ObjectPool.reportStats = function reportStats() {
+    // Iterate over pools, report their current size
+
+    var i;
+    for (i = 0; i < ObjectPool.pools.length; i++) {
```

Comment:
`ObjectPool` can't be bootstrapped without a `statsd` because of the assertions, so the timer will never be set and this function will never be called if there's no `statsd`. The check was actually unnecessary.